### PR TITLE
Bounce attack tests

### PR DESCRIPTION
**Pull request by terencechain**
#3989 was merged before we had a chance to push the tests. This PR implements unit tests for bounce attack resistance as implemented in #3989

### CODE REVIEW

File: beacon-chain/blockchain/forkchoice/process_block.go [hunk 1]
```diff
@@ -486,7 +486,7 @@ func (s *Store) currentSlot() uint64 {
 
 // updates justified check point in store if a better check point is known
 func (s *Store) updateJustifiedCheckpoint() {
-	if helpers.SlotsSinceEpochStarts(s.currentSlot()) == 0 {
+	if !(helpers.SlotsSinceEpochStarts(s.currentSlot()) == 0) {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/forkchoice/process_block.go]:
```suggestion
	if helpers.SlotsSinceEpochStarts(s.currentSlot()) > 0 {
```
this is more clear imo

Comment (prestonvanloon) [reply beacon-chain/blockchain/forkchoice/process_block.go]: Wait is this what you want?

Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/process_block.go]: No, I dont think that's it. Here's the spec reference:
```python
if not (compute_slots_since_epoch_start(current_slot) == 0):
...
```
Only proceed if it's the first slot of the epoch. I'm going to simplify it to
```go
if s.currentSlot() % params.beaconConfig.SlotsPerSlot == 0 {
}
```

Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/process_block.go]: Updated to
```go
	if s.currentSlot() % params.BeaconConfig().SlotsPerEpoch != 0 {
		return
	}
```

Comment (prestonvanloon) [reply beacon-chain/blockchain/forkchoice/process_block.go]: Can you add a comment? It's not clear of the intention. Is this "if slot is an epoch start slot, don't update the justified checkpoint"? 

If so, why?

Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/process_block.go]: The intension is to "only check and update best justified epoch to last justified epoch" at the epoch boundary slot. Think of epoch transition, it only happens once an epoch

Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/process_block.go]: Added a comment

File: beacon-chain/blockchain/forkchoice/process_block.go [hunk 2]
```diff
@@ -486,7 +486,8 @@ func (s *Store) currentSlot() uint64 {
 
 // updates justified check point in store if a better check point is known
 func (s *Store) updateJustifiedCheckpoint() {
-	if helpers.SlotsSinceEpochStarts(s.currentSlot()) == 0 {
+	// only need to perform this update at epoch boundary slot
+	if s.currentSlot() % params.BeaconConfig().SlotsPerEpoch != 0 {
```
Comment (rauljordan) [inline beacon-chain/blockchain/forkchoice/process_block.go]: Don't we have a helper for this?

File: beacon-chain/blockchain/forkchoice/process_block.go [hunk 3]
```diff
@@ -486,7 +486,8 @@ func (s *Store) currentSlot() uint64 {
 
 // updates justified check point in store if a better check point is known
 func (s *Store) updateJustifiedCheckpoint() {
-	if helpers.SlotsSinceEpochStarts(s.currentSlot()) == 0 {
+	// Update at epoch boundary slot only
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/forkchoice/process_block.go]: Thanks

**Pull request merged**